gtsam/geometry/SO4.cpp

```diff
@@ -0,0 +1,226 @@
+
+/**
+ * @file    SO4.cpp
+ * @brief   4*4 matrix representation of SO(4)
+ * @author  Frank Dellaert
+ * @author  Luca Carlone
+ */
+
+#include <gtsam/base/concepts.h>
+#include <gtsam/base/timing.h>
+#include <gtsam/geometry/SO4.h>
+#include <gtsam/geometry/Unit3.h>
+
+#include <Eigen/Eigenvalues>
+#include <boost/random.hpp>
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+namespace gtsam {
+
+// /* *************************************************************************
+// */ static Vector3 randomOmega(boost::mt19937 &rng) {
+//   static boost::uniform_real<double> randomAngle(-M_PI, M_PI);
+//   return Unit3::Random(rng).unitVector() * randomAngle(rng);
+// }
+
+// /* *************************************************************************
+// */
+// // Create random SO(4) element using direct product of lie algebras.
+// SO4 SO4::Random(boost::mt19937 &rng) {
+//   Vector6 delta;
+//   delta << randomOmega(rng), randomOmega(rng);
+//   return SO4::Expmap(delta);
+// }
+
+//******************************************************************************
+template <>
+Matrix4 SO4::Hat(const Vector6& xi) {
+  // skew symmetric matrix X = xi^
+  // Unlike Luca, makes upper-left the SO(3) subgroup.
+  Matrix4 Y = Z_4x4;
+  Y(0, 1) = -xi(5);
+  Y(0, 2) = +xi(4);
+  Y(1, 2) = -xi(3);
+  Y(0, 3) = -xi(2);
+  Y(1, 3) = +xi(1);
+  Y(2, 3) = -xi(0);
+  return Y - Y.transpose();
+}
+
+//******************************************************************************
+template <>
+Vector6 SO4::Vee(const Matrix4& X) {
+  Vector6 xi;
+  xi(5) = -X(0, 1);
+  xi(4) = +X(0, 2);
+  xi(3) = -X(1, 2);
+  xi(2) = -X(0, 3);
+  xi(1) = +X(1, 3);
+  xi(0) = -X(2, 3);
+  return xi;
+}
+
+//******************************************************************************
+/* Exponential map, porting MATLAB implementation by Luca, which follows
+ * "SOME REMARKS ON THE EXPONENTIAL MAP ON THE GROUPS SO(n) AND SE(n)" by
+ * Ramona-Andreaa Rohan */
+template <>
+SO4 SO4::Expmap(const Vector6& xi, ChartJacobian H) {
+  using namespace std;
+  if (H) throw std::runtime_error("SO4::Expmap Jacobian");
+
+  // skew symmetric matrix X = xi^
+  const Matrix4 X = Hat(xi);
+
+  // do eigen-decomposition
+  auto eig = Eigen::EigenSolver<Matrix4>(X);
+  Eigen::Vector4cd e = eig.eigenvalues();
+  using std::abs;
+  sort(e.data(), e.data() + 4, [](complex<double> a, complex<double> b) {
+    return abs(a.imag()) > abs(b.imag());
+  });
+
+  // Get a and b from eigenvalues +/i ai and +/- bi
+  double a = e[0].imag(), b = e[2].imag();
+  if (!e.real().isZero() || e[1].imag() != -a || e[3].imag() != -b) {
+    throw runtime_error("SO4::Expmap: wrong eigenvalues.");
+  }
+
+  // Build expX = exp(xi^)
+  Matrix4 expX;
+  using std::cos;
+  using std::sin;
+  const auto X2 = X * X;
+  const auto X3 = X2 * X;
+  double a2 = a * a, a3 = a2 * a, b2 = b * b, b3 = b2 * b;
+  if (a != 0 && b == 0) {
```

Comment:
should this be `abs(b) < eps`?

Reply:
Yeah...

Reply:
Then the problem becomes what should `eps` be...

Reply:
just std::epsilon ?

Reply:
I wonder if they are big enough for our application?

I have made a small test snippet: https://godbolt.org/z/rX5F4E
```
a = 2.22045e-16, c2 = 0, c3 = 0
a = 2.22045e-15, c2 = 0, c3 = 0
a = 2.22045e-14, c2 = 0, c3 = 0
a = 2.22045e-13, c2 = 0, c3 = 0
a = 2.22045e-12, c2 = 0, c3 = 0
a = 2.22045e-11, c2 = 0, c3 = 0
a = 2.22045e-10, c2 = 0, c3 = 0
a = 2.22045e-09, c2 = 0, c3 = 0
a = 2.22045e-08, c2 = 0.45036, c3 = 0.302231
a = 2.22045e-07, c2 = 0.4999, c3 = 0.166832
a = 2.22045e-06, c2 = 0.49999, c3 = 0.166658
a = 2.22045e-05, c2 = 0.5, c3 = 0.166667
a = 0.000222045, c2 = 0.5, c3 = 0.166667
a = 0.00222045, c2 = 0.5, c3 = 0.166667
a = 0.0222045, c2 = 0.499979, c3 = 0.166663
```

Reply:
I tested this against a Taylor series impl, I think Taylor series is much more accurate when the number is small: https://godbolt.org/z/pASumE

```matlab
==== TESTING a!=0 && b == 0 ====
a = eps*10^0,   c2 = 0.0000000000,   c3 = 0.0000000000
a = eps*10^0, c2_s = 0.5000000000, c3_s = 0.1666666667
a = eps*10^1,   c2 = 0.0000000000,   c3 = 0.0000000000
a = eps*10^1, c2_s = 0.5000000000, c3_s = 0.1666666667
a = eps*10^2,   c2 = 0.0000000000,   c3 = 0.0000000000
a = eps*10^2, c2_s = 0.5000000000, c3_s = 0.1666666667
a = eps*10^3,   c2 = 0.0000000000,   c3 = 0.0000000000
a = eps*10^3, c2_s = 0.5000000000, c3_s = 0.1666666667
a = eps*10^4,   c2 = 0.0000000000,   c3 = 0.0000000000
a = eps*10^4, c2_s = 0.5000000000, c3_s = 0.1666666667
a = eps*10^5,   c2 = 0.0000000000,   c3 = 0.0000000000
a = eps*10^5, c2_s = 0.5000000000, c3_s = 0.1666666667
a = eps*10^6,   c2 = 0.0000000000,   c3 = 0.0000000000
a = eps*10^6, c2_s = 0.5000000000, c3_s = 0.1666666667
a = eps*10^7,   c2 = 0.0000000000,   c3 = 0.0000000000
a = eps*10^7, c2_s = 0.5000000000, c3_s = 0.1666666667
a = eps*10^8,   c2 = 0.0000000000,   c3 = 0.0000000000
a = eps*10^8, c2_s = 0.5000000000, c3_s = 0.1666666667
a = eps*10^9,   c2 = 0.0000000000,   c3 = 0.0000000000
a = eps*10^9, c2_s = 0.5000000000, c3_s = 0.1666666667
a = eps*10^10,   c2 = 0.5072854620,   c3 = 0.1584563250
a = eps*10^10, c2_s = 0.5000000000, c3_s = 0.1666666667
a = eps*10^11,   c2 = 0.4999067644,   c3 = 0.1666960539
a = eps*10^11, c2_s = 0.4999999995, c3_s = 0.1666666667
a = eps*10^12,   c2 = 0.4999989981,   c3 = 0.1666656303
a = eps*10^12, c2_s = 0.4999999955, c3_s = 0.1666666667
a = eps*10^13,   c2 = 0.5000000127,   c3 = 0.1666666444
a = eps*10^13, c2_s = 0.4999999548, c3_s = 0.1666666667
a = eps*10^14,   c2 = 0.4999999998,   c3 = 0.1666666665
a = eps*10^14, c2_s = 0.4999995482, c3_s = 0.1666666667
a = eps*10^15,   c2 = 0.4999999995,   c3 = 0.1666666666
a = eps*10^15, c2_s = 0.4999954825, c3_s = 0.1666666666
a = eps*10^16,   c2 = 0.4999999510,   c3 = 0.1666666569
a = eps*10^16, c2_s = 0.4999548249, c3_s = 0.1666666569
a = eps*10^17,   c2 = 0.4999951021,   c3 = 0.1666656871
a = eps*10^17, c2_s = 0.4995482491, c3_s = 0.1666656871
a = eps*10^18,   c2 = 0.4995104026,   c3 = 0.1665687362
a = eps*10^18, c2_s = 0.4954826828, c3_s = 0.1665687362
a = eps*10^19,   c2 = 0.4529004540,   c3 = 0.1571406159
a = eps*10^19, c2_s = 0.4567037733, c3_s = 0.1571405684
a = eps*10^20,   c2 = 0.0098082126,   c3 = 0.0092824648
a = eps*10^20, c2_s = -21.0450562957, c3_s = -2.5473590444 # diverge
```

Reply:
@dellaert Should we use a Taylor series based implementation for smaller `a` and `b`s? The results of the original implementation seems to be wrong if `a` is smaller than `3*10^-5`

Reply:
@ProfFan this is great stuff! However, could we do the Taylor expansion into a smaller, more targeted SO4 that institutionalizes some of your tests above into unit tests? I really want to land this PR. I would love to see the Taylor expansion improve on the existing (Luca) implementation though.

Reply:
@dellaert I think the PR is ready to land in its current state, as all tests now pass and we have no conflicts. I can prepare another PR for SO4 afterwards.